### ISSUE TITLE
TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received type object

### ISSUE THREAD

**Issue by Mathieu-R** [GoogleChromeLabs/prerender-loader#3]
Hello !

I tried to prerender my preact application but I get this error.
If I remove `prerender-loader` and just put the path to my index.html I don't get any error.
`Module build failed: TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received type object`

I'm using `html-webpack-plugin` like so : 
```js 
new htmlWebpackPlugin({
  // prerendering with prerender-loader
  template: `!!prerender-loader?string!${path.resolve(__dirname, '../src/index.html')}`,
  minify: {
    removeComments: true
  },
  cache: true,
  // make it work consistently with multiple chunks
  chunksSortMode: 'dependency'
})
```

Thanks for the help ! 😃

**Comment by rkichenama** [comment]
Looks like in `dist/prerender-loader.js:91`, parentCompiler.options.entry is an object reflecting the `config.entry` object

**Comment by Mathieu-R** [comment]
Yes, I had on object in config.entry, removed it and no problem now.

**Comment by developit** [comment]
Whoops - I should have caught that!  It would be nice if `prerender-loader` took that into account and used the last entry of the Array or the `main` property of an object.

Or, perhaps the best solution would be to check for multi-entry configurations and respect them by using `MultiEntryPlugin` here:
https://github.com/GoogleChromeLabs/prerender-loader/blob/5f667781679351b8f160a0e39e0deb45099d2848/src/index.js#L128

**Comment by cvetanov** [comment]
Hello,

First of all kudos to the idea and the actual implementation and thanks for publishing it!
I was about to post a new issue and then I saw this one so here it goes.

I tried it with create-react-app and got the following (similar) error:
```
TypeError: Path must be a string. Received [ 'C:\\data\\playground\\playground-js\\config\\polyfills.js',  
  - 'C:\\data\\playground\\playground-js\\src\\index.js' ]
```

The webpack configuration's entry is an actual array which besides the main bundle, also includes polyfills. Now, this exact use case might not be that relevant since there is a low possibility that the prerenderd content will use polyfills, but anyhow, what do you think about supporting multiple entry points?

A workaround would be importing the polyfills (in this case), or any other entries, into one single entry and using only that one, but it would be nice to have support for multiple entries making the project structure more modular (not coupling polyfills with some library imports and the actual application in order to be compliant with `prerender-loader`).

The error is originating from this [line](https://github.com/GoogleChromeLabs/prerender-loader/blob/master/src/index.js#L95) where `path.relative(context, parentCompiler.options.entry)` gets executed and throws the error. Would it be possible to add conditionals which check the type of the entry object before processing it?

I might be way off target here since I have very little knowledge of the webpack internals (the 'compilers'). Is this even possible? If yes, I would be glad to dig in and learn stuff along the way.

**Comment by tizano** [comment]
Hello,
I have the same issue as @cvetanov 
If someone could help us or write an example. That will be great 😄

**Comment by cvetanov** [comment]
Hi @tizano, a workaround would be to create a new file (let's call it `entry.js`) where you will import all of the files from the array of entries and use that file as a single entry.
Let's say your entries are made up of `['file1.js', 'file2.js', 'index.js']`, then your webpack config would change to `entry: 'entry.js'` and your entry.js file would import all of the files (mind the relative paths):
```
import './file1.js';
import './file2.js';
import './index.js';
```

My comment was more of a question if I should go ahead and dig in the implementation and add support for array/object-like entries?

**Comment by MrEfrem** [comment]
I use conditions logic for build a `entry` array for a different environment (debug, dev, prod), therefore using the single file impossible.

**Comment by developit** [comment]
It should be fairly straightforward to account for multi-entry bundles and have them work out-of-the-box in prerender-loader.

**Comment by CarrierDirectRoss** [comment]
I am receiving the same error `Module build failed: TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received type object`

with a config.entry of
`entry: './main.jsx',`

and a loader configuration of
```
new HtmlWebPackPlugin({
    template: `!!prerender-loader?string!${path.resolve(__dirname, '../src/assets/index.html')}`,
    filename: 'index.html',
    favicon: './assets/favicon.ico',
    inject: false,
})
```

**Comment by developit** [comment]
@CarrierDirectRoss are you sure there isn't a plugin manipulating your config?  Things like HotModuleReplacementPlugin (or the `--hot` flag) will convert `entry` to an Array and add items to it.

**Comment by injms** [comment]
I get this error in a setup with only `webpack` and `HtmlWebPackPlugin` - so no other Webpack plugins are manipulating the config. It's a clean install of everything (aside from Node and NPM) at their latest versions on WSL. (This isn't anything that hasn't been said before, but another point of data might be useful.)

If there's a workaround, would someone be able to give me a hint please? 😄

**Comment by vaxilicaihouxian** [comment]
same issue.
any one can fix this?

**Comment by vaxilicaihouxian** [comment]
@injms Did you use webpack-dev-server? I met the same issue.And I found 'path.relative(xxx,entry)' that entry is an array.I only use webpack-dev-server and html-webpack-plugin.

here is my config file 

```
var path = require('path');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const webpack = require('webpack');
module.exports = {
    context:__dirname,
    entry: 'src/app.js',
    output:{
        filename:'bundle.js',
        path:path.resolve(__dirname,'dist')
    },
    plugins: [
        new HtmlWebpackPlugin({
            template:'!!prerender-loader?string!src/index.html'
            
        })
    ],
    // module:{
    //     rules:[
    //         {
    //             test: path.resolve(__dirname,'src/index.html'),
    //             loader:'prerender-loader?string'
    //         }
    //     ]
    // },
    devServer:{
        // contentBase:'./dist',
         hot:false
    },
    mode:'development'
}
```

**Comment by injms** [comment]
@vaxilicaihouxian - yep, I'm using `webpack-dev-server`. If I run `webpack` to build then this will happily render the page, so I'm currently using Nodemon and BrowserSync for my hot-reloading needs.

**Comment by vaxilicaihouxian** [comment]
@injms I try to change *path.relative(xxx,entry)* to *path.relative(xxx,entry[x])* to fix this only for my development.

**Comment by developit** [comment]
Fixed in [1.1.0](https://github.com/GoogleChromeLabs/prerender-loader/releases/tag/1.1.0). Thanks for all the help!

**Comment by nschubach** [comment]
1.1.0 is still giving me this error:

```
ERROR in   Error: Child compilation failed:
  Module build failed (from ../node_modules/prerender-loader/dist/prerender-loader.js):
  TypeError: Path must be a string. Received [ '<snip>/node_modules/webpack-dev-server/client/index.js?http://localhost:8080',
    '<snip>/node_modules/webpack/hot/dev-server.js',
    'src' ]
```

```
$ npm show prerender-loader version
1.1.0
```

I currently just have the one named entry point (the root):
```{ '.': 'src' }```
